src/cargo/util/logger.rs

```diff
@@ -0,0 +1,110 @@
+//! Build analysis logging infrastructure.
+
+use std::io::{BufWriter, Write};
+use std::mem::ManuallyDrop;
+use std::path::Path;
+use std::sync::mpsc;
+use std::sync::mpsc::Sender;
+use std::thread::JoinHandle;
+
+use cargo_util::paths;
+
+use crate::CargoResult;
+use crate::core::Workspace;
+use crate::util::log_message::LogMessage;
+use crate::util::short_hash;
+
+/// Logger for `-Zbuild-analysis`.
+pub struct BuildLogger {
+    tx: ManuallyDrop<Sender<String>>,
+    run_id: String,
+    handle: Option<JoinHandle<()>>,
+}
+
+impl BuildLogger {
+    /// Creates a logger if `-Zbuild-analysis` is enabled.
+    pub fn maybe_new(ws: &Workspace<'_>) -> CargoResult<Option<Self>> {
+        let analysis = ws.gctx().build_config()?.analysis.as_ref();
+        match (analysis, ws.gctx().cli_unstable().build_analysis) {
+            (Some(analysis), true) if analysis.enabled => Ok(Some(Self::new(ws)?)),
+            (Some(_), false) => {
+                ws.gctx().shell().warn(
+                    "ignoring 'build.analysis' config, pass `-Zbuild-analysis` to enable it",
+                )?;
+                Ok(None)
+            }
+            _ => Ok(None),
+        }
+    }
+
+    fn new(ws: &Workspace<'_>) -> CargoResult<Self> {
+        let run_id = Self::generate_run_id(ws)?;
+
+        let log_dir = ws.gctx().home().join("build-analysis");
+        paths::create_dir_all(log_dir.as_path_unlocked())?;
+
+        let filename = format!("{run_id}.jsonl");
+        let log_file = log_dir.open_rw_exclusive_create(
+            Path::new(&filename),
+            ws.gctx(),
+            "build analysis log",
+        )?;
+
+        let (tx, rx) = mpsc::channel::<String>();
+
+        let handle = std::thread::spawn(move || {
+            let mut writer = BufWriter::new(log_file);
+            for json_line in rx {
+                let _ = writer.write_all(json_line.as_bytes());
+                let _ = writer.write_all(b"\n");
+            }
+            let _ = writer.flush();
+        });
+
+        Ok(Self {
+            tx: ManuallyDrop::new(tx),
+            run_id,
+            handle: Some(handle),
+        })
+    }
+
+    /// Generates a unique run ID.
+    ///
+    /// The format is `{hash}-{timestamp}`, with `:` and `.` in the timestamp
+    /// replaced by `-` to make it safe for filenames.
+    /// For example, `f891d525d52ecab3-2025-10-24T19-45-02-773638Z`.
+    pub fn generate_run_id(ws: &Workspace<'_>) -> CargoResult<String> {
+        let real_path = std::fs::canonicalize(ws.root())?;
+        let hash = short_hash(&real_path);
+        let timestamp = jiff::Timestamp::now()
+            .to_string()
+            .replace(':', "-")
+            .replace('.', "-");
+        Ok(format!("{hash}-{timestamp}"))
```

Comment:
Time stamp should likely go first for easier scanning and/or log rotation

Reply:
Fixed